src/localisation/include/EKF.h

```diff
@@ -0,0 +1,166 @@
+                .finished()};
+    // covariance of the encoder speed measurement and the imu angular velocity
+    // measurement
+    Matrix2d pu{
+    (Matrix2d() << sdev_speed * sdev_speed, 0, 0, sdev_gyro* sdev_gyro)
+    .finished()};
+    // observation matrix
+    Matrix3d h{(Matrix3d() << 1, 0, 0, 0, 1, 0, 0, 0, 1).finished()};
+    // covariance of the gps position measurements and the imu orientation
+    // measurement
+    Matrix3d r{(Matrix3d() << sdev_gps_x_measurement * sdev_gps_x_measurement,
+                0,
+                0,
+                0,
+                sdev_gps_y_measurement* sdev_gps_y_measurement,
+                0,
+                0,
+                0,
+                sdev_angle_measurement* sdev_angle_measurement)
+               .finished()};
+    // covariance for initial position currently assume it's perfect
+    Matrix3d intial_p{(Matrix3d() << 0, 0, 0, 0, 0, 0, 0, 0, 0).finished()};
+
+    /**
+     * Callback function for when new GPS data is received
+     *
+     * @param The GPS data received in the callback in the form of an odometry
+     * msg
+     */
+    void gpsCallBack(const nav_msgs::Odometry::ConstPtr& gps_message);
+
+    /**
+     * Callback function for when new encoder data is received
+     *
+     * @param The encoder data received in the callback in the form of an
+     * odometry msg
+     */
+    void encoderCallBack(const nav_msgs::Odometry::ConstPtr& encoder_message);
+
+    /**
+     * Callback function for when new IMU data is received
+     *
+     * @param The IMU data received in the callback in the form of an IMU msg
+     */
+    void imuCallBack(const sensor_msgs::Imu::ConstPtr& imu_message);
+
+    /**
+     * Publishes the estimated postion calculated from the EKF
+     *
+     * @param Takes in a geometry_msgs/Pose message
+     */
+    void publishPose(geometry_msgs::Pose pose_msg);
+
+  public:
+    // Constructor
+    EKF(int argc, char** argv, std::string node_name);
+
+    /**
+     * Takes an angle and adjusts it if necessary so that it says within the
+     * bounds of -pi and +pi currently an angle that equals to pi, -pi, 3pi,
+     * -3pi, etc will be rebound to pi
+     *
+     * @param an angle in radians
+     *
+     * @return a rebounded angle in radians
+     */
+    static double defineAngleInBounds(double angle);
+
+    /**
+     * Takes in an angle of orientation on the z axis and converts it to
+     * quaternion form
+     *
+     * @param angle of orientation on the z axis in radians
+     *
+     * @return angle in quaternion form
+     */
+    static geometry_msgs::Quaternion angleToQuaternion(double angle);
+
+    /**
+     * Takes in angle of orientation in quaternion form and converts it to
+     * an angle of orientation on the z axis
+     *
+     * @param angle in quaternion form
+     *
+     * @return angle of orientation on the z axis in radians
+     */
+    static double quaternionToAngle(geometry_msgs::Quaternion quat_angle);
```

Comment:
We should probably be using the `tf::getYaw` function that ROS provides to get our yaw from a quaternion. In this case, the less code we have to worry about testing and verifying, the better. You should be able to get `tf::getYaw` by including `transform_datatypes.h`. The doc is [here](http://docs.ros.org/api/tf/html/c++/namespacetf.html#a33e93f413622e296d0a2a93b252bbb4b)